### DISCUSSION TITLE
SimplePlanner interpretation of cartesian waypoint seed

### DISCUSSION THREAD

**Discussion by rjoomen** [tesseract-robotics/tesseract_planning#375] [Q&A]
In the simple planner, the seed of a cartesian waypoint [is used directly](https://github.com/tesseract-robotics/tesseract_planning/blob/master/tesseract_motion_planners/simple/src/interpolation.cpp#L217). This way, it seems to me a cartesian waypoint with seed waypoint effectively functions as a joint waypoint: the simple planner interpolates towards the seed, instead of the cartesian waypoint.

I was expecting the seed to be used as seed for the IK, like so:
```
j2 = getClosestJointSolution(base, base_cwp.getSeed().position);
```
Could you explain the reasoning behind this? And how can I set an actual seed for a cartesian waypoint that does produce a correctly interpolated path?

**Comment by Levi-Armstrong** [comment]
The simple planner in most cases is to generate a seed trajectory which is then passed to the planners. If the cartesian waypoint already has one it is assumed this is the seed you want it too use and it will avoid running IK. Though I do see your use case and I think there are two option, the first create a new set of profiles to have this behavior or add an option to the existing profiles to change how seeds are handled if provided for cartesian waypoints. I would vote for the latter option.

**Reply by marip8** [reply]
> If the cartesian waypoint already has one it is assumed this is the seed you want it too use and it will avoid running IK.

To me this is not the definition of a seed. To me a seed is either the start condition of a gradient-based IK solver, or something used to bias the choice between a set of IK solutions - not something that should implicitly be considered an actual valid IK solution. I think a planner should always have to solve IK for a Cartesian waypoint and could optionally use the provided seed to do so (which is @rjoomen's suggestion and I think your latter option).

Another option would be to clarify what is meant by `seed` by renaming it to something more specific. Does it mean a seed for solving IK (which is my current assumption)? Or does it mean a seed for a new joint waypoint (which is the assumption from the current behavior)?

**Reply by Levi-Armstrong** [reply]
> To me this is not the definition of a seed. 

This depend on the context in which you are viewing it from. The seed was added to the cartesian waypoints to allow representing the seed trajectory within the same data structure where before we had two different data structures which made things complicated. This is the reason for the default behavior being: If a seed is already provided for a waypoint that does not implicitly have one then it is used within the simple planner. If you are viewing it from the IK perspective then the seed is as described, but that was not the reason behind adding this to the cartesian waypoint.

**Reply by Levi-Armstrong** [reply]
> Another option would be to clarify what is meant by seed by renaming it to something more specific. Does it mean a seed for solving IK (which is my current assumption)? Or does it mean a seed for a new joint waypoint (which is the assumption from the current behavior)?

I am not sure this will help at the cartesian waypoint definition because it is up to the profiles to decide how they want to interpret the data. We can at least document it at the profile if it is not already.

**Reply by marip8** [reply]
> I am not sure this will help at the cartesian waypoint definition because it is up to the profiles to decide how they want to interpret the data.

I see your point, but currently the profiles technically don't make this decision. They actually use a stand-alone function (linked in original post) to interpolate the Cartesian waypoint. This function makes that assumption that the Cartesian waypoint should be replaced by the seed and doesn't have the option to use the seed for IK instead.

Why wouldn't you just replace a Cartesian waypoint with a joint waypoint from the beginning if the intent is to replace it with a joint seed position?

**Reply by Levi-Armstrong** [reply]
> I see your point, but currently the profiles technically don't make this decision. 

The simple planner LVS profile should be interpolating based on linear or freespace type and then the seed assigned should use the provided seed state if one exists and if not it will run IK.

``` c++
  if (base_cwp.hasSeed())
    j2 = base_cwp.getSeed().position;
  else
    j2 = getClosestJointSolution(base, j1);
```

>Why wouldn't you just replace a Cartesian waypoint with a joint waypoint from the beginning if the intent is to replace it with a joint seed position?

The cartesian information need to be preserved for planners like trajopt which can represent cartesian waypoints within the planner and the seed associated would be used to initialize the solver state.

**Reply by marip8** [reply]
The function you linked is for interpolation, but it doesn't truly interpolate between the two input waypoints. It copies a seed instruction (which is not necessarily a valid IK solution for the given Cartesian waypoint) and interpolates between a "fake" joint waypoint and the target joint waypoint. My point is that this function shouldn't be making this decision, the profile should be

> The cartesian information need to be preserved for planners like trajopt which can represent cartesian waypoints within the planner and the seed associated would be used to initialize the solver state.

Yes, but why not just solve IK? I guess it's more flexible technically, but at the end of the day won't you still be using IK at some level to generate this seed?

If you're going to use the seed in this way, I think it should be explicitly differentiated from an IK seed. Maybe it should be called `instruction_seed` or `joint_state_approximation` with a docstring that describes its intended purpose

**Reply by Levi-Armstrong** [reply]
> Yes, but why not just solve IK? I guess it's more flexible technically, but at the end of the day won't you still be using IK at some level to generate this seed?

This breaks chaining planners together because the seed would get over written within the planner. Also IK can be expensive so I wanted to avoid unnecessary calls throughout the pipeline. This way if a seed exists it is used and no IK is ran.

**Reply by Levi-Armstrong** [reply]
>The function you linked is for interpolation, but it doesn't truly interpolate between the two input waypoints. It copies a seed instruction (which is not necessarily a valid IK solution for the given Cartesian waypoint) and interpolates between a "fake" joint waypoint and the target joint waypoint. My point is that this function shouldn't be making this decision, the profile should be

The functions were explicitly written for the simple planner profiles and are separated functions to make the code more readable and easier to test. So the profiles are making the decision. As I suggested we could updated the existing profiles to allow this different behavior which would then result in the functions being updated to support the option.

**Reply by Levi-Armstrong** [reply]
>If you're going to use the seed in this way, I think it should be explicitly differentiated from an IK seed. Maybe it should be called instruction_seed or joint_state_approximation with a docstring that describes its intended purpose

The way the seed is used is not dictated by the waypoint but by the profiles, but renaming the method to getJointStateApprox() and setJointStateApprox() may cause less confusion.

**Comment by rjoomen** [comment]
Upon giving this a bit more thought I realized that probably when setting a seed I might as well do a full IK and use that as the seed. But I'll keep the idea of adding this as an option in mind.

**Reply by Levi-Armstrong** [reply]
>Upon giving this a bit more thought I realized that probably when setting a seed I might as well do a full IK and use that as the seed. But I'll keep the idea of adding this as an option in mind.

This is typically what I do because a cartesian waypoint may have multiple solution and I want to be explicit and make it use a specific one.
